### ISSUE TITLE
PyTorch Implementation

### ISSUE THREAD

**Issue by koustav123** [subpic/ava-mlsp#4]
I am trying to implement the  3FC head in PyTorch and train it from scratch using the wide features extracted using TF. I trained the TF model from scratch with these features and the results were reproduced. But in PyTorch the model fails to learn well.  Here's the model

```
class Model_Pool_3FC_2(nn.Module):
    def __init__(self, args, dset_classes, in_features = 3072, fc_x = 2048, feature_width = 5):
        super(Model_Pool_3FC_2, self).__init__()
        self.inception = Inception_34(16928,1024)

        self.cap_3FC = nn.Sequential(OrderedDict([
            ('final_GAP', nn.AvgPool2d(feature_width)),
            ('flatten', ResNet_Base.Flatten()),

            ('head_fc0', nn.Linear(in_features, fc_x)),
            ('head_act0', nn.ReLU()),
            ('head_bn0', Custom_BN_layer(1, fc_x)),
            ('head_do0', nn.Dropout(0.25)),

            ('head_fc1', nn.Linear(fc_x, int(fc_x / 2))),
            ('head_act1', nn.ReLU()),
            ('head_bn1', Custom_BN_layer(1, int(fc_x / 2))),
            ('head_do1', nn.Dropout(0.25)),

            ('head_fc2', nn.Linear(int(fc_x / 2), int(fc_x / 8))),
            ('head_act2', nn.ReLU()),
            ('head_bn2', Custom_BN_layer(1, int(fc_x / 8))),
            ('head_do2', nn.Dropout(0.5)),

            ('head_out', nn.Linear(int(fc_x / 8), len(dset_classes)))

            ]))


    def forward(self, x):
        x = x.cuda()
        inc_o = self.inception(x)
        y_A2 = self.cap_3FC(inc_o)
        return {'A2': y_A2}


    def set_training_mode(self, phase):
        if phase == 'train':
            self.inception.train(True)
            self.cap_3FC.train(True)
        else:
            self.eval()`

```
And Here's the Inception block
```
class Inception_34(nn.Module):
    def __init__(self, in_features, out_features):
        super(Inception_34, self).__init__()
        #pdb.set_trace()
        self.branch_1x1 = BasicConv2d(in_features, out_features,'branch_1x1', kernel_size=1, stride=1)

        self.branch_33_11 = BasicConv2d(in_features, out_features, '3x3_1x1', kernel_size=1, stride=1)
        self.branch_33_13 = BasicConv2d(out_features, int(out_features/2), '3x3_1x3',kernel_size=(1,3), stride=1, padding = (0,1))
        self.branch_33_31 = BasicConv2d(out_features, int(out_features/2),'3x3_3x1' ,kernel_size=(3,1), stride=1, padding = (1,0))

        self.avg_pool_2D = nn.AvgPool2d(kernel_size=3, stride=1, padding=1)
        self.branch_pool = BasicConv2d(in_features, out_features, 'branch_pool',kernel_size=1, stride=1)


    def forward(self, x):
        o_11 = self.branch_1x1(x)
        o_33_11 = self.branch_33_11(x)
        o_33 = torch.cat((self.branch_33_13(o_33_11),self.branch_33_31(o_33_11)), 1)
        o_pool = self.branch_pool(self.avg_pool_2D(x))
        return torch.cat((o_11, o_33, o_pool), 1)
`
```
And here are the BasicConv blocks and a BN layer
```
class BasicConv2d(nn.Module):

    def __init__(self, in_planes,  out_planes,  name, kernel_size, stride=1, padding=0):
        super(BasicConv2d, self).__init__()
        self.net = nn.Sequential(OrderedDict([
            (name +'_conv', nn.Conv2d(int(in_planes), int(out_planes),
                              kernel_size=kernel_size, stride=stride,
                              padding=padding, bias=False)),
            (name +'_bn', Custom_BN_layer(2, int(out_planes))),
            (name +'_relu', nn.ReLU(inplace=False))
        ]))

    def forward(self, x):
        return self.net(x)

class Custom_BN_layer(nn.Module):
    def __init__(self, dim, in_feat, affine=True, eps = 1e-5, momentum = 0.1):
        super(Custom_BN_layer, self).__init__()

        if dim == 1:
            self.bn = nn.BatchNorm1d(in_feat, eps = eps, affine=affine, momentum=momentum)
        else:
            self.bn = nn.BatchNorm2d(in_feat, eps = eps, affine=affine, momentum=momentum)

    def forward(self, x):
        return (x)

```
The other details for training I tried  are as follows:
LR : 1e-4 , 1e-5, 1e-3
Loss : MSE with 1 -D output
Optimizer: Tried both ADAM/SGD
Dropout: As specified in the model and also tried removing dropout
Precision: Both 32 and 16

The problem is the network does not learn well. The loss decreases but then plateaus after 2/3 epochs. The prediction saturates around 5.4, which is the average score. The accuracy and balanced accuracy is 70 and 50, respectively ie random. 

I verified the labels and they are okay. I am not being able to figure out why the network is learning so badly. 
Here's the training(red) and validation(grey) loss

![A2_Loss_MSE_](https://user-images.githubusercontent.com/3684879/74086103-a94a3d80-4a77-11ea-9bb9-a90af5f32f91.png)

Here's the predicted MOS for validation after  saturation
![MOS_V](https://user-images.githubusercontent.com/3684879/74086002-c0d4f680-4a76-11ea-80fc-3e16c671d522.png)
And here's the one for randomly sam[pled training data
![MOS_T](https://user-images.githubusercontent.com/3684879/74086011-d518f380-4a76-11ea-81fe-08ec070a79ac.png)

Any suggestions? Thanks in advance

**Comment by subpic** [comment]
I could only guess what the problem is. You could try loading the features in PyTorch directly rather than using TF; maybe there's a bug there? The correspondence between features and scores might have gotten scrambled, leading to apparently training on random labels with the same distribution?
